### PR TITLE
Better listeners structure

### LINKED ISSUE #82: Garbage Collection
It seems that the VirtualCollection instances will not be garbage collected, since for each accepted model an event listener is registered:
```
_rebuildIndex: function () {
...
    this.collection.each(function (model, i) {
      if (this.accepts(model, i)) {
        model.on('all', this._onAllEvent, this);
...
}
```
The event listener is however not deregistered inside the ```bindLifecycle``` method. I guess ```model.off(...)``` should be called when a destroy event is emitted, e.g.:
```
bindLifecycle: function (view, method_name) {
  view.on(method_name, _.bind(function() {
    this.stopListening();
    _.invoke(this.models, 'off', 'all', this._onAllEvent, this);
  }, this));
}
```
Any thoughts?

### PR DESCRIPTION
**Pull request by p3drosola**
fixes #82 

@dtodor what do you think?

### CODE REVIEW

**Comment by dtodor** [comment]
Looks good to me.

**Comment by megawac** [comment]
Whats the advantage of this over the simple fix of `model.on` -> `this.listenTo`?

**Comment by disruptek** [comment]
Can we move towards listening on fewer events, and doing so explicitly, versus listening on `all`?  Listening on the individual events reduces wasted calls for events not associated with collection manipulation.  Explicit event specifications are easier to find and follow in the code.

**Comment by megawac** [comment]
I agree. Also easier for step through debugging purposes

File: backbone.virtual-collection.js [hunk 1]
```diff
@@ -62,6 +53,30 @@ var VirtualCollection = Backbone.VirtualCollection = Backbone.Collection.extend(
...
+    } else if (target instanceof Backbone.Model) {
+      switch (event) {
+        case 'add':
+          return this._onAdd.apply(this, args);
+        case 'change':
+          return this._onChange.apply(this, args);
+        case 'remove':
+          return this._onRemove.apply(this, args);
+      }
+      return this.trigger.apply(this, args_with_name);
```
Comment (megawac) [inline backbone.virtual-collection.js]: we didn't use to proxy all events on the parent collection. is this really desired

File: backbone.virtual-collection.js [hunk 2]
```diff
@@ -62,6 +53,30 @@ var VirtualCollection = Backbone.VirtualCollection = Backbone.Collection.extend(
...
+    var args_with_name = _.toArray(arguments);
+    var args = args_with_name.slice(1);
+    if (target == this.collection) {
+      switch (event) {
+        case 'reset':
+          return this._onReset.apply(this, args);
+        case 'sort':
+          return this._onSort.apply(this, args);
+      }
+      return this.trigger.apply(this, args_with_name);
```
Comment (mllocs) [inline backbone.virtual-collection.js]: yep, same here

**Comment by p3drosola** [comment]
Yeah, I'm aware about the change in proxying all the other events. To me it makes sense because the idea is for a virtualcollection to be functionally equivalent to a regular one, unless there's a good reason for it not to be.

I realize this is change in behaviour though. So it could cause some problems.

About the listener structure. I find this is simpler because there's one listener bound. and one unbound. That way we're sure not to leave anything dangling. I think the logic is easier to follow as well. It's all in one method, instead of spread all around. You don't need to keep a bunch of criss-crossing handlers (some on the models, some on the collection) in your head to see what would happen.

@disruptek can you expound on your concerns?

@mllocs what do you think?

**Comment by mllocs** [comment]
@p3drosola I think that using the `all` event instead of the specific ones may change the behavior. 

It looks like Backbone uses the `all` event internally and does some manipulation before sending the specific event.

**Comment by megawac** [comment]
Yeah, don't it also capture all change:* events, even for models outside
the virtual collection

**Comment by disruptek** [comment]
It really doesn't matter how simple the code for VirtualCollection is, though it's already a simple module.  Our goal here is to present the best possible experience to the developers and users.  Those who develop VC are exclusive of that goal -- we shouldn't be looking for ways to simplify the VC code at the expense of others.

In addition to the basic style concerns associated with using `all` versus individual event names, the change creates a more monolithic interface against Backbone, in which we cannot surgically add or remove listeners for events.  This might not feel like a problem at the moment, but consider how we might try to work around a future interoperability problem with this "enhancement".

The argument for making VC behave identically to Backbone falls flat because we can't control Backbone's interface and VC wouldn't exist if its collection behavior was identical.  As long as we need to play catch-up to Backbone's interface and perform magic of our own before exposing our own interface, I don't see how attempting to force VC to behave identically is anything less than a downgrade.  Backbone is a starting point, not an ending point; it's a platform from which to extend and expand functionality.

For example, it might be ideal if VC absorbed `sort` events that don't change the order of the collection, or presented alternate 'sort-that-didn't-change-order' and 'sort-that-did-change-order' events.  I'd like to see an 'added-this-model-at-index-X' event as well, which could prevent gratuitous strikes by the crude `sort` hammer.  And I'd like to see VC->VC interfaces made more efficient than VC->Collection interfaces.  These ideas are orthogonal to homogenizing the VC interface to that of Collection, and clearly these features are easier for a developer user to implement in their own extensions of VC without the monolithic `all` event handler.

In my case, I don't limit collection events to only those utilized by VirtualCollection, so this will reduce performance of very UX-sensitive code in every case where it's unnecessary.  I'm probably an exception, but I have extensive chains of VC->VC->VC projections which operate on dynamic collections of thousands of items.  Maybe I'm lazy, but I don't look forward to maintaining a less serviceable and extensible VC just to buy performance I already have today.

**Comment by megawac** [comment]
Another option would be to use the new and improved `"update"` event (https://github.com/jashkenas/backbone/pull/3948) being added in `v1.2.4` which will list the `added` and `removed` models, as well as merged models (`change` events on models currently not included here but feel free to comment on Backbone).

This will allow essentially all of the collection listeners to be merged and then you can just rebroadcast watched model events

**Comment by p3drosola** [comment]
Ok, makes sense to me. Thanks for weighing in guys.

**Pull request closed without merging**